### ISSUE TITLE
Ankidroid - Anki 2.1 - Mathjax

### ISSUE THREAD

**Issue by p3l4h0** [ankidroid/Anki-Android#4915]
###### Research
*Enter an [ x ] character to confirm the points below:*

[X] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[X] I have searched for similar existing issues here and on the user forum

###### Reproduction Steps

1. start anki desktop create mathjax card /( 1+1 /)
2. on anki desktop ist beautiful 1+1 
3. on ankidroid still /( 1+1 /)

###### Expected Result

it should make inline formula like old latex pictures only with vectorgraphics

###### Actual Result

only code

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

**Comment by mikehardy** [comment]
Here is the relevant part of the manual: https://apps.ankiweb.net/docs/manual.html#mathjax-support

Here's standard use of MathJax in Android:
https://github.com/mathjax/MathJax-docs/wiki/Guide:-MathJax-in-Android-apps

...but here's the upstream version in Python (and JS/CSS for MathJax) as a base for any port:
https://github.com/dae/anki/tree/master/web/mathjax

Integrating it just enough to display cards correctly during review would be the natural first step.

As a side note, I didn't see anything else besides the V2 scheduler (already noted in #4905) in v2.1+ but I could be missing something

**Comment by mikehardy** [comment]
If anyone has a good sample deck (especially with clozures in it) that would be helpful to link directly or attach, as edge cases may be difficult

**Comment by tsudoko** [comment]
> As a side note, I didn't see anything else besides the V2 scheduler (already noted in #4905) in v2.1+ but I could be missing something

There are minor changes to the sync protocol regarding deletions, but the old way seems to be still supported for now: https://github.com/dae/anki/commit/eb44584b290181e0397a231e6291ca36b86d4189

**Comment by mikehardy** [comment]
@tsudoko - thanks for the pointer!

**Comment by Moosehorse123** [comment]
> If anyone has a good sample deck (especially with clozures in it) that would be helpful to link directly or attach, as edge cases may be difficult

Do you still need a sample deck?

**Comment by mikehardy** [comment]
@Moosehorse123  Yes!

**Comment by Moosehorse123** [comment]
> @Moosehorse123 Yes!

Ok! I have plenty deck options, let me know what specifications that you require other than cloze cards. Also does your listed email address work to send it to?

**Comment by mikehardy** [comment]
I don't have exact specifications, just something that would be good as a verifier to make sure everything worked. So if there are mathjax expressions that are rare + difficult, then those. And integrating cloze and mathjax (with overlaps if Anki Desktop supports it) is important to make sure we handle that correctly - it looked like the hardest part of the Anki Desktop implementation. Other than that, just not too big - shouldn't need more than a few cards I think? And my preference would be to have it attached here so anyone could use it if they started work. Thanks!

**Comment by alxmjo** [comment]
I was able to get a workaround going with Anki 2.0 (on Mac desktop) and `$` and `$$` as delimiters. This worked properly on AnkiDroid (though it was a bit slow to render). This was the template I used:

```
{{cloze:Text}}
<script type="text/x-mathjax-config">
    MathJax.Hub.processSectionDelay = 0;
    MathJax.Hub.Config({
        messageStyle:"none",
        showProcessingMessages:false,
        tex2jax:{
            inlineMath: [['$','$']],
            displayMath:[['$$','$$']],
            processEscapes:true
        }
});
</script>
<script type="text/javascript">
(function() {
  if (window.MathJax != null) {
    var card = document.querySelector('.card');
    MathJax.Hub.Queue(['Typeset', MathJax.Hub, card]);
    return;
  }
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_SVG-full';
  document.body.appendChild(script);
})();
</script>
``` 

Now that I've started using 2.1 (along with the built-in delimiters `\(  \)` and \[  \]), MathJax is no longer rendered. I tried updating the note template with the new delimiters, but it didn't seem to have an effect.

Is there any workaround for this right now?

**Comment by alxmjo** [comment]
As for the sample deck, here's one of mine with MathJax, Cloze deletions, basic cards, etc.

https://www.dropbox.com/s/epw3qbg8limlmip/LinearAlgebraDeck.apkg?dl=0

**Comment by mikehardy** [comment]
Unfortunately no workaround that I know of but I'm pretty impressed you managed it with 2.0 - that's a nice bit of code. It's my understanding that all the AbstractFlashcardReviewer/TemplateEditor code would need a look for this to be complete, but that simply including the MathJax distribution in Anki and referencing it in an rendering hook [like LaTeX did](https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/libanki/LaTeX.java) might be enough to get rendering working

Thanks very much for the deck, even though memories of linear algebra class make me shudder :-). Not my finest class. I'm attaching it directly here so it doesn't disappear

[LinearAlgebraDeck.apkg.zip](https://github.com/ankidroid/Anki-Android/files/2514985/LinearAlgebraDeck.apkg.zip)

**Comment by alxmjo** [comment]
Should've just zipped it, thanks. 

That LaTeX class just displays LaTeX images, correct? For how Anki used to handle LaTeX?

I'd love to dig into this but should probably try my hand at a simpler item first. It's been a long time since I've done anything with Java/Android.

As for the workaround, I feel like if it were working under 2.0 then getting it going with 2.1 shouldn't be too onerous.

**Comment by mikehardy** [comment]
> That LaTeX class just displays LaTeX images, correct? For how Anki used to handle LaTeX?

True, but it does this relatively simple task through the external render hooks infrastructure, which if I'm understanding correctly gives the full HTML contents of the card to each hook in turn and lets them do whatever they want with it. So you could look for the new delimiters via regex and rewrite it with whatever hook was needed to reference the MathJax package (which should be in-app / local, IMHO but could be referenced by CDN like you did to start)

> I'd love to dig into this but should probably try my hand at a simpler item first. It's been a long time since I've done anything with Java/Android.

I think a full-fledged implementation (like where AnkiDesktop doesn't highlight cloze when mathjax is present) is pretty tightly coupled and so difficult, but giving it a whirl with an external hook might actually be a quick win. Or at least a time-boxed test :-)

> As for the workaround, I feel like if it were working under 2.0 then getting it going with 2.1 shouldn't be too onerous.

Because all the heavy-lifting is in Javascript we would just bundle (or reference, as you did) while the on-device webview does the work, I agree it should be *almost* trivial.

**Comment by alxmjo** [comment]
Thanks for the comments, @mikehardy.

I just managed to get MathJax, Anki 2.1, and AnkiDroid all playing nicely together. The key was adding an extra backslash in my note template, like so:

```
{{cloze:Text}}
<script type="text/x-mathjax-config">
    MathJax.Hub.processSectionDelay = 0;
    MathJax.Hub.Config({
        messageStyle:"none",
        showProcessingMessages:false,
        tex2jax:{
            inlineMath: [ ['$','$'], ['\\(','\\)'] ],
            displayMath: [ ['$$','$$'], ['\[','\]'] ],
            processEscapes:true
        }
});
</script>
<script type="text/javascript">
(function() {
  if (window.MathJax != null) {
    var card = document.querySelector('.card');
    MathJax.Hub.Queue(['Typeset', MathJax.Hub, card]);
    return;
  }
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_SVG-full';
  document.body.appendChild(script);
})();
</script>
```

Now LaTeX displays correctly in both Anki 2.1 and AnkiDroid.

Cloze deletions are still less than perfect (they take up an entire block before expansion instead of appearing inline - see attachments) but that's off topic for this thread.

![screenshot_20181025-082733](https://user-images.githubusercontent.com/1656106/47518238-3f386580-d83f-11e8-94a8-8f9a21d1728d.png)
![screenshot_20181025-082739](https://user-images.githubusercontent.com/1656106/47518239-3f386580-d83f-11e8-864c-6551185f1aed.png)

**Comment by mikehardy** [comment]
Hey, that's fantastic and thanks for sharing the code so others can use it until it's built-in.

**Comment by alxmjo** [comment]
My pleasure. Hopefully I can find some time to dig into this myself.

**Comment by maxhebditch** [comment]
> Thanks for the comments, @mikehardy.
> 
> I just managed to get MathJax, Anki 2.1, and AnkiDroid all playing nicely together. The key was adding an extra backslash in my note template, like so:
> 
> ```
> {{cloze:Text}}
> <script type="text/x-mathjax-config">
>     MathJax.Hub.processSectionDelay = 0;
>     MathJax.Hub.Config({
>         messageStyle:"none",
>         showProcessingMessages:false,
>         tex2jax:{
>             inlineMath: [ ['$','$'], ['\\(','\\)'] ],
>             displayMath: [ ['$$','$$'], ['\\[','\\]'] ],
>             processEscapes:true
>         }
> });
> </script>
> <script type="text/javascript">
> (function() {
>   if (window.MathJax != null) {
>     var card = document.querySelector('.card');
>     MathJax.Hub.Queue(['Typeset', MathJax.Hub, card]);
>     return;
>   }
>   var script = document.createElement('script');
>   script.type = 'text/javascript';
>   script.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_SVG-full';
>   document.body.appendChild(script);
> })();
> </script>
> ```
> 
> Now LaTeX displays correctly in both Anki 2.1 and AnkiDroid.

This works great! Only issue if you have some maths that naturally includes `}}` because that it gets interpreted as a cloze close.

e.g.
````
{{c1::\(\beta\)}}
{{c2::\(\beta = \frac{\alpha}{\sum_{0}^{1}}\)}}
````

will give you

front
![image](https://user-images.githubusercontent.com/4867412/49294035-b77ef000-f4a9-11e8-98af-5fed6e02573c.png)
back
![image](https://user-images.githubusercontent.com/4867412/49294105-eb5a1580-f4a9-11e8-9499-781d345e55b2.png)

but if I had a space, going from `{1}}` to `{1} }` I get

front
![image](https://user-images.githubusercontent.com/4867412/49294094-da110900-f4a9-11e8-87e2-331da19ff935.png)
back
![image](https://user-images.githubusercontent.com/4867412/49294098-dda49000-f4a9-11e8-8948-c8d56d2d4213.png)

**Comment by p3l4h0** [comment]
Thanks @alxmjo and @mikehardy with after reinstalling the apkg on the firetablet it works.

**Comment by jaspercpes** [comment]
> ```
> {{cloze:Text}}
> <script type="text/x-mathjax-config">
>     MathJax.Hub.processSectionDelay = 0;
>     MathJax.Hub.Config({
>         messageStyle:"none",
>         showProcessingMessages:false,
>         tex2jax:{
>             inlineMath: [ ['$','$'], ['\\(','\\)'] ],
>             displayMath: [ ['$$','$$'], ['\\[','\\]'] ],
>             processEscapes:true
>         }
> });
> </script>
> <script type="text/javascript">
> (function() {
>   if (window.MathJax != null) {
>     var card = document.querySelector('.card');
>     MathJax.Hub.Queue(['Typeset', MathJax.Hub, card]);
>     return;
>   }
>   var script = document.createElement('script');
>   script.type = 'text/javascript';
>   script.src = 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_SVG-full';
>   document.body.appendChild(script);
> })();
> </script>
> ```
> 
> Now LaTeX displays correctly in both Anki 2.1 and AnkiDroid.

Am I missing something or does this no longer work?
I'm using this template for my cloze cards and it displays fine in Ankidroid, but on desktop (version 2.1.5 on Linux), it does not do anything and instead displays the source text, i.e. including the dollar signs. Are there any necessary settings I overlooked?

**Comment by colinpretorius** [comment]
@jaspercpes I don't think the dollar signs are meant to be supported on the desktop client - the [help docs](https://apps.ankiweb.net/docs/manual.html#mathjax-support) mention that only `\(, \)` and `\[, \]` are supported, and these work for me on the desktop client and AnkiDroid with this workaround.

**Comment by colinpretorius** [comment]
As just mentioned, this workaround works for me - I've been using it for a while now and the rendering is much nicer than with LaTeX. 

A few points worth mentioning:
* rendering is slower - an opened note will briefly show the markup before re-rendering the MathJax. I haven't seen this happen with the desktop client, so I presume some manner of pre-rendering happens there.
* wide LaTeX image files get scaled down to fit the AnkiDroid screen - MathJax rendering goes off the side. I don't know whether this would be a bug or a feature :-) but the scaling was convenient, whereas I now have to be careful when creating cards to avoid having to scroll left/right.
* perhaps most importantly from a compatibility point of view - with this workaround you can use MathJax inside a cloze deletion, but a cloze deletion inside a MathJax block will 'break' the rendering and just show the raw markup. The Anki desktop client renders this correctly with MathJax (and doing the same with LaTex worked fine, too).

**Comment by mikehardy** [comment]
@colinpretorius thanks for the feedback! We obviously hope to support mathjax natively but until then this is the only way to do it I'm aware of. Hopefully your info helps all that need mathjax until we get it Built-In

**Comment by p3l4h0** [comment]
thanks to you i can now learn nice formulas on my e-book reader:
![anki](https://user-images.githubusercontent.com/35264331/53576874-ec681500-3b74-11e9-9d95-125d7eda7b42.jpg)

**Comment by alxmjo** [comment]
@p3l4h0 Cool! Which e-reader is that?

**Comment by adamwolf** [comment]
Hi folks!  Is this being worked on, and if not, does the approach planned in this thread still seem like a good idea?

**Comment by mikehardy** [comment]
@adamwolf hello! No, there is no current effort in progress for Mathjax - the javascript-based workaround is the only thing in action now. I outlined the full upstream change as a basis for a full implementation in AnkiDroid here https://github.com/ankidroid/Anki-Android/issues/4915#issuecomment-416692525 (above) and that was my best guess at the correct implementation at the time, I don't think anything should have changed.

**Comment by Vistian** [comment]
Any ideas on how can we get Ankidroid to play nicely with mhchem?

**Comment by adamwolf** [comment]
I made some basic progress with this.  There are some tricky things to be done with MathJax and clozes that I need to port tests over for, and see if I need to help make them pass.

The desktop client actually starts with the cards set invisible, and fades them in quickly after MathJax finishes the rendering of the math parts.  I have recreated that as well.  I'm certain there will be strong opinions on whether we should only do that on MathJax cards, or on all cards.  I do not care either way, and it's easy to change.

@mikehardy (and anyone else appropriate) would you want me to copy in the MathJax stuff as a submodule or by just copying the files in and labeling what version they are or something else?

**Comment by mikehardy** [comment]
Awesome! And yes, the cloze looked the worst part of it

haha - people will *scream* if all cards fade. They may even prefer to do a placeholder + mathjax render unless the fade can be terminated as soon as mathjax is provably rendered. The userbase is very sensitive to even small lag on card review.

I believe Mathjax can be updated independently yes? Like, we want to have it local to save bandwidth etc, but it is conceivable the whole package - which comes from an outside source - might bump versions without us doing anything, just like a gradle dependency?

If I don't have that right please correct me, but if I do then I would like them as separate as possible, but in as close to their upstream format as possible, so that we can manage updates on the dependency easily.

**Comment by adamwolf** [comment]
You are correct regarding Mathjax upgrading without us having to do anything except a git pull and recompile.  I will include instructions in a README alongside the external dependency.

Another option is to eliminate the fade completely. On non-Mathjax cards, keep it 100% exactly the same, and on Mathjax cards, snap the text in the moment Mathjax finishes.  I suppose we could set it up so that the behavior is configurable via javascript variables, which users could set in their templates.  In fact, this may be how desktop Anki is already set up.  I will investigate.

I even have mhchem working.

<img width="330" alt="Screenshot 2019-04-18 09 15 52" src="https://user-images.githubusercontent.com/35681/56367539-0a3c1880-61bb-11e9-8baf-30978556afd6.png">

**Comment by mikehardy** [comment]
Oh that looks really nice! This is a popular feature request, it is going to help a lot of people I think.

**Comment by adamwolf** [comment]
Not to get offtopic, but I also need video stuff, so I'll be looking at that stuff next.

**Comment by mikehardy** [comment]
If you wanted to tackle the one everyone asks for, you could get #4756 in shape to merge - that one actually works, it's just moving the working functionality from the PR to a different style, or maybe @timrae will let this one in just for now to clear the queue pending an "advanced note editor" rework in the future. And not sure what you mean by video stuff but I'll always try to help if it is some bit of the code I know about. Cheers

**Comment by mikehardy** [comment]
Hey everyone! Go test the new releases I made here: https://github.com/mikehardy/Anki-Android/releases/tag/v2.9alpha69

They are parallel releases so they won't mess up your main install. But you can sync AnkiWeb or load apkg files or use them normally otherwise. The 2.9alpha69 release I just made has the current PR in it and I want some testing on it.

We understand the compatibility method described here does not work, but raw mathjax does work. If people could confirm that, and describe how difficult it is to revert the compatibility thing, I'd appreciate it. I would love to hear that it works, reverting the stopgap is easy, and we can merge this thing.

Cheers

**Comment by adamwolf** [comment]
@mikehardy So I installed your preview, and tried *my* cards on it.  My MathJax cards work fine on this preview, even with the fix.  The Anki test suite cards work fine.  The Linear Algebra test deck renders fine, but it is not using the built-in Mathjax.

I'm feeling great about this.    

Folks should remove the Compatibility stuff so that their Ankidroid isn't grabbing the external Mathjax all the time.

**Comment by mikehardy** [comment]
Well I'm just going to close this here issue right now ;-). It's merged! It's live! 

Go [grab your new alpha version](https://github.com/ankidroid/Anki-Android/wiki/Installation#beta--alpha-releases) folks, and stop using compatibility templates

*Please* raise new issues if you see them, but otherwise every time you use Mathjax in AnkiDroid spare a nice thought for @adamwolf - nice work, sir
